node/src/tasks/bsp_charge_fees.rs

```diff
@@ -50,6 +82,68 @@ where
             storage_hub_handler,
         }
     }
+
+    async fn remove_file_from_forest(&self, file_key: &H256) -> anyhow::Result<()> {
+        // Remove the file key from the Forest.
+        // Check that the new Forest root matches the one on-chain.
+        {
+            let fs = self
+                .storage_hub_handler
+                .forest_storage_handler
+                .get(&NoKey)
+                .await
+                .ok_or_else(|| anyhow!("Failed to get forest storage."))?;
+
+            fs.write().await.delete_file_key(file_key).map_err(|e| {
+                error!(target: LOG_TARGET, "CRITICAL❗️❗️ Failed to apply mutation to Forest storage. This may result in a mismatch between the Forest root on-chain and in this node. \nThis is a critical bug. Please report it to the StorageHub team. \nError: {:?}", e);
+                anyhow!(
+                    "Failed to remove file key from Forest storage: {:?}",
+                    e
+                )
+            })?;
+        };
+
+        Ok(())
+    }
+
+    async fn check_provider_root(&self, provider_id: ProviderId) -> anyhow::Result<()> {
+        // Get root for this provider according to the runtime.
+        let onchain_root = self
+            .storage_hub_handler
+            .blockchain
+            .query_provider_forest_root(provider_id)
+            .await
+            .map_err(|e| {
+                error!(target: LOG_TARGET, "CRITICAL❗️❗️ Failed to query provider root from runtime after successfully submitting proof. This may result in a mismatch between the Forest root on-chain and in this node. \nThis is a critical bug. Please report it to the StorageHub team. \nError: {:?}", e);
+                anyhow!(
+                    "Failed to query provider root from runtime after successfully submitting proof: {:?}",
+                    e
+                )
+            })?;
+
+        trace!(target: LOG_TARGET, "Provider root according to runtime: {:?}", onchain_root);
+
+        // Check that the new Forest root matches the one on-chain.
+        let fs = self
+            .storage_hub_handler
+            .forest_storage_handler
+            .get(&NoKey)
+            .await
+            .ok_or_else(|| anyhow!("Failed to get forest storage."))?;
+
+        let root = { fs.read().await.root() };
+
+        trace!(target: LOG_TARGET, "Provider root according to Forest Storage: {:?}", root);
+
+        if root != onchain_root {
+            error!(target: LOG_TARGET, "CRITICAL❗️❗️ Applying mutations yielded different root than the one on-chain. This means that there is a mismatch between the Forest root on-chain and in this node. \nThis is a critical bug. Please report it to the StorageHub team.");
```

Comment:
We return an error that then gets propagated up through the result from `handle_event`, I believe that should be sufficient, no? @ffarall does the same thing in the `submit_proof` tasks so if we modify this one we might have to modify both

Reply:
The difference is that if you miss a submit proof you can still catch up, if you have a root mismatch where/how do you sort it out?

Reply:
Yeah but it's not for missing to submit a proof, it's actually exactly the same code: after submitting a valid proof, it checks that the forest root matches the one on-chain, and errors out if it doesn't. It's not something you can recover from though, and it should never happen, that's why the error states to contact the StorageHub team since it's an inconsistency error

Reply:
I think the error log is the right way to go. We're signalling strongly that this is a critical bug, but we're not panicking. Even if this is unrecoverable, the worst that can happen is that from here on the forest-related transactions will fail, until someone manually does something (most likely). But, by keeping the node alive, you can maybe query it or investigate further why this happened.